content/development/testing.md

```diff
@@ -9,7 +9,312 @@ chapter: false
+```
+
+Note, that the default value does not have the MIME type `application/json`.
+
+In your `crs-setup.conf` you need to add these extra rules (after the rule `900990`):
+
+```
+SecAction
+    "id:900005,\
+    phase:1,\
+    nolog,\
+    pass,\
+    ctl:ruleEngine=DetectionOnly,\
+    ctl:ruleRemoveById=910000,\
+    setvar:tx.blocking_paranoia_level=4,\
+    setvar:tx.crs_validate_utf8_encoding=1,\
+    setvar:tx.arg_name_length=100,\
+    setvar:tx.arg_length=400,\
+    setvar:tx.max_file_size=64100,\
+    setvar:tx.combined_file_sizes=65535"
+
+SecRule REQUEST_HEADERS:X-CRS-Test "@rx ^.*$" \
+    "id:999999,\
+    phase:1,\
+    pass,\
+    t:none,\
+    log,\
+    msg:'%{MATCHED_VAR}'"
+```
+
+Now, after the restarting the web server all request will be sent to `httpbin`. Let's start testing.
+
+## Go-ftw
+
+Tests are performed using [go-ftw](https://github.com/coreruleset/go-ftw). We run our test suite automatically using **go-ftw** as part of a [GitHub workflow](https://github.com/coreruleset/coreruleset/blob/{{< param crs_dev_branch >}}/.github/workflows/test.yml). You can easily reproduce that locally, on your workstation.
+
+For that you will need:
+
+- the CRS Git repository
+- Docker (modern versions of docker already include the `compose` command, if you are running an older version you also need to have `docker-compose` installed)
+  OR
+  your own environment (see above installation steps)
+- your rules and tests!
+
+You can download pre-compiled binaries of **go-ftw** or build from source (requires you to have a **Go** environment). The pre-compiled binaries are available [on GitHub](https://github.com/coreruleset/go-ftw/releases). The binaries are ready to run and do not require installation.
+
+You can also install pre-compiled binaries by using `go install`, if you have a **Go** environment:
+
+```bash
+$ go install github.com/coreruleset/go-ftw@latest
+```
+
+This will install the binary into your `$HOME/go/bin` directory. To compile **go-ftw** from source, run the following commands:
+
+```bash
+$ git clone https://github.com/coreruleset/go-ftw.git
+$ cd go-ftw
+$ go build .
+```
+
+This will build the binary in the **go-ftw** repository.
+
+Now create a configuration file. Because Apache httpd and nginx use different log file paths, and, perhaps, different ports, you may want to create two different configuration files for **go-ftw**.
+
+Example for nginx:
+
+```bash
+$ cat .ftw.nginx.yaml
+logfile: /var/log/nginx/error.log
+logmarkerheadername: X-CRS-TEST
+testoverride:
+  input:
+    dest_addr: "127.0.0.1"
+    port: 8080
+mode: "default"
+```
+
+Example for Apache httpd:
+
+```bash
+$ cat .ftw.apache.yaml
+logfile: /var/log/apache2/error.log
+logmarkerheadername: X-CRS-TEST
+testoverride:
+  input:
+    dest_addr: "127.0.0.1"
+    port: 80
+mode: "default"
+```
+
+Please verify that these settings are correct for your setup, specifically the `port` values.
```

Comment:
Why repeating the default config values? Mode is `"default"` by... default. Same with the `logmarkerheadername`.

Reply:
Because I see that in the official [go-ftw documentation](https://github.com/coreruleset/go-ftw#yaml-config-file)

Reply:
Ok. Then we might want to point people to that documentation instead? I think that if we repeat things all over the place, they will be out of sync sooner than later...

Reply:
I agree we might show the original documentation, but **beside** this and not **instead**. I would like to see the whole preparation process in one page. What do you think?

Reply:
But what would be the solution?
* put the original documentation **before** the examples
* remove `mode: "default"` settings
* both of them

Reply:
I would:
- link the upstream doc before the examples
- remove everything from the config except the bare minimal for tests to work

Reply:
Done. Could you check it?